README.md
# VUI-DEMO
一个vue UI 组件库构建框架

## 安装依赖
```
npm run init
```

### 运行示例
```
npm run dev
```

### 打包示例demo
```
npm run build
```

### 代码校验
```
npm run lint
```

### 打包生成 ui 组件库文件
```
npm run lib
```

### 生成一个组件模板框架
```
npm run new:comp
```
### 删除一个组件模板框架
```
npm run del:comp
```

### 规范提交 git commit 根据提示输入相关日志内柔
```
npm run commit
```

### 整个生成 changelog.md 
```
npm run initlog
```
### 追加 changelog.md 文档
```
npm run addlog
```

### Customize configuration
See [Configuration Reference](https://cli.vuejs.org/config/).


## 目录结构说明

examples —— 是组件示例展示demo界面
packages —— 组件库核心代码
lib —— 组件库打包后的代码
scripts—— 工具文件——生成组件、删除组件 以及模板文件


## 参考
1. [使用 Vue cli 3.0 构建自定义组件库的方法](https://www.yisu.com/zixun/165977.html)
2. [规范代码 git 提交格式](https://www.cnblogs.com/mengfangui/p/12634845.html)


CHANGELOG.md
# 0.1.0 (2022-03-24)


### Bug Fixes

* **package.json:** changelog 文档 c9f9034


### Features

* **build:** readme 文档更新 c0f24ba





scripts/template.js
/*
 * @Author: koko
 * @Date: 2022-03-23 17:14:03
 * @FilePath: \san-player\scripts\template.js
 * @LastEditTime: 2022-03-24 09:52:47
 * @LastEditors: koko
 * @Description: 是生成组件代码的模板文件
 */
module.exports = {
  vueTemplate: (compoenntName,cssName) => {
    compoenntName =
      compoenntName.charAt(0).toLowerCase() + compoenntName.slice(1);
    return `<!--
    * @Author: koko
    * @Date: 2022-03-24 09:42:55
    * @LastEditTime: 2022-03-24 09:43:13
    * @LastEditors: koko
    * @Description: 
   -->
   
   <template>
     <div class="st-${cssName}">${compoenntName}</div>
   </template>
   <script>
   export default {
     name: "${compoenntName}",
     data() {
       return {};
     },
     props: {},
     methods: {},
   };
   </script>
   <style lang="scss" scope>
   .st-${cssName} {
   }
   </style>
   `;
  },
  entryTemplate: (compoenntName) => {
    return `/*
    * @Author: koko
    * @Date: 2022-03-24 09:39:29
    * @LastEditTime: 2022-03-24 09:40:09
    * @LastEditors: koko
    * @Description: ${compoenntName}
    */
   
   import ${compoenntName} from "./${compoenntName}";
   
   ${compoenntName}.install = function (Vue) {
     Vue.component(${compoenntName}.name, ${compoenntName});
   };
   
   if (typeof window !== "undefined" && window.Vue) {
     window.Vue.component(${compoenntName}.name, ${compoenntName});
   }
   `;
  },
  mdDocs: (title) => {
    return `# ${title} 组件
    <!-- {.md} -->
    ---
    <!-- {.md} -->
    ## 如何使用
    <!-- {.md} -->
    ## Attributes
    <!-- {.md} -->
    | 参数 | 说明 | 类型 | 可选值 | 默认值 |
    |-----|-----|-----|-----|-----|
    | - | - | - | - | - |`;
  },
};


scripts/delete-comp.js
/*
 * @Author: koko
 * @Date: 2022-03-23 17:13:45
 * @FilePath: \san-player\scripts\delete-comp.js
 * @LastEditTime: 2022-03-23 17:15:16
 * @LastEditors: koko
 * @Description: 是用来删除无用的组件目录和自定义组件说明文档脚本
 */
// 删除自定义组件脚本
const chalk = require("chalk");
const path = require("path");
const fs = require("fs-extra");
const uppercamelize = require("uppercamelcase");
const resolve = (...file) => path.resolve(__dirname, ...file);
const log = (message) => console.log(chalk.green(`${message}`));
const successLog = (message) => console.log(chalk.blue(`${message}`));
const errorLog = (error) => console.log(chalk.red(`${error}`));
log("请输入要删除的组件名称,形如 demo 或者 demo-test");
process.stdin.on("data", async (chunk) => {
  let inputName = String(chunk).trim().toString();
  inputName = uppercamelize(inputName);
  const componentDirectory = resolve("../packages", inputName);
  const hasComponentDirectory = fs.existsSync(componentDirectory);
  const docsDirectory = resolve("../examples/docs");
  const docsMdName = resolve(docsDirectory, `${inputName}.md`);
  if (inputName) {
    if (hasComponentDirectory) {
      log(`删除 component 目录 ${componentDirectory}`);
      await removePromise(componentDirectory);
      successLog(`已删除 ${inputName} 组件目录`);
      log(`删除 component 文档 ${docsMdName}`);
      fs.unlink(docsMdName);
      successLog(`已删除 ${inputName} 组件说明文档`);
    } else {
      errorLog(`${inputName}组件目录不存在`);
      return;
    }
  } else {
    errorLog(`请重新输入组件名称:`);
    return;
  }
  process.stdin.emit("end");
});
process.stdin.on("end", () => {
  log("exit");
  process.exit();
});
function removePromise(dir) {
  return new Promise(function (resolve, reject) {
    // 先读文件夹
    fs.stat(dir, function (_err, stat) {
      if (stat.isDirectory()) {
        fs.readdir(dir, function (_err, files) {
          files = files.map((file) => path.join(dir, file)); // a/b a/m
          files = files.map((file) => removePromise(file)); // 这时候变成了promise
          Promise.all(files).then(function () {
            fs.rmdir(dir, resolve);
          });
        });
      } else {
        fs.unlink(dir, resolve);
      }
    });
  });
}


scripts/create-comp.js
/*
 * @Author: koko
 * @Date: 2022-03-23 17:13:23
 * @FilePath: \san-player\scripts\create-comp.js
 * @LastEditTime: 2022-03-24 09:48:29
 * @LastEditors: koko
 * @Description: 是用来生成自定义组件目录和自定义组件说明文档脚本
 */
// 创建自定义组件脚本
const chalk = require("chalk");
const path = require("path");
const fs = require("fs-extra");
const uppercamelize = require("uppercamelcase");
const resolve = (...file) => path.resolve(__dirname, ...file);
const log = (message) => console.log(chalk.green(`${message}`));
const successLog = (message) => console.log(chalk.blue(`${message}`));
const errorLog = (error) => console.log(chalk.red(`${error}`));
const { vueTemplate, entryTemplate, mdDocs } = require("./template");
const generateFile = (path, data) => {
  if (fs.existsSync(path)) {
    errorLog(`${path}文件已存在`);
    return;
  }
  return new Promise((resolve, reject) => {
    fs.writeFile(path, data, "utf8", (err) => {
      if (err) {
        errorLog(err.message);
        reject(err);
      } else {
        resolve(true);
      }
    });
  });
};
// 这里生成自定义组件
log("请输入要生成的组件名称,形如 demo 或者 demo-test");
let componentName = "";
process.stdin.on("data", async (chunk) => {

  // 组件名称
  let cssName = String(chunk).trim().toString();
  inputName = uppercamelize(cssName);

  const componentDirectory = resolve("../packages", inputName);
  const componentVueName = resolve(componentDirectory, `${inputName}.vue`);
  const entryComponentName = resolve(componentDirectory, "index.js");
  const hasComponentDirectory = fs.existsSync(componentDirectory);
  if (inputName) {
    // 这里生成组件
    if (hasComponentDirectory) {
      errorLog(`${inputName}组件目录已存在，请重新输入`);
      return;
    } else {
      log(`生成 component 目录 ${componentDirectory}`);
      await dotExistDirectoryCreate(componentDirectory);
    }
    try {
      if (inputName.includes("/")) {
        const inputArr = inputName.split("/");
        componentName = inputArr[inputArr.length - 1];
      } else {
        componentName = inputName;
      }
      log(`生成 vue 文件 ${componentVueName}`);
      await generateFile(componentVueName, vueTemplate(componentName,cssName));
      log(`生成 entry 文件 ${entryComponentName}`);
      await generateFile(entryComponentName, entryTemplate(componentName));
      successLog("生成 component 成功");
    } catch (e) {
      errorLog(e.message);
    }
  } else {
    errorLog(`请重新输入组件名称:`);
    return;
  }
  // 这里生成自定义组件说明文档
  const docsDirectory = resolve("../examples/docs");
  const docsMdName = resolve(docsDirectory, `${inputName}.md`);
  try {
    log(`生成 component 文档 ${docsMdName}`);
    await generateFile(docsMdName, mdDocs(`${inputName} 组件`));
    successLog("生成 component 文档成功");
  } catch (e) {
    errorLog(e.message);
  }
  process.stdin.emit("end");
});
process.stdin.on("end", () => {
  log("exit");
  process.exit();
});
function dotExistDirectoryCreate(directory) {
  return new Promise((resolve) => {
    mkdirs(directory, function () {
      resolve(true);
    });
  });
}
// 递归创建目录
function mkdirs(directory, callback) {
  var exists = fs.existsSync(directory);
  if (exists) {
    callback();
  } else {
    mkdirs(path.dirname(directory), function () {
      fs.mkdirSync(directory);
      callback();
    });
  }
}
